lib/node_modules/@stdlib/ndarray/flatten-by/lib/main.js

```diff
@@ -0,0 +1,106 @@
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var isPlainObject = require( '@stdlib/assert/is-plain-object' );
+var isFunction = require( '@stdlib/assert/is-function' );
+var isndarrayLike = require( '@stdlib/assert/is-ndarray-like' );
+var map = require( '@stdlib/ndarray/map' );
+var flatten = require( '@stdlib/ndarray/flatten' );
+var format = require( '@stdlib/string/format' );
+
+
+// MAIN //
+
+/**
+* Applies a callback to elements in an input ndarray and returns a flattened output ndarray.
+*
+* @param {ndarray} x - input ndarray
+* @param {Options} [options] - function options
+* @param {NonNegativeInteger} [options.depth] - maximum number of dimensions to flatten
+* @param {string} [options.order='row-major'] - order in which input ndarray elements should be flattened
+* @param {Function} fcn - callback function
+* @param {*} thisArg - callback execution context
+* @throws {TypeError} first argument must be an ndarray-like object
+* @throws {TypeError} options argument must be an object
+* @throws {TypeError} callback argument must be a function
+* @throws {TypeError} must provide valid options
+* @returns {ndarray} output ndarray
+*
+* @example
+* var array = require( '@stdlib/ndarray/array' );
+* var ndarray2array = require( '@stdlib/ndarray/to-array' );
+*
+* function scale( value ) {
+*     return value * 2.0;
+* }
+*
+* var x = array( [ [ [ 1.0, 2.0 ] ], [ [ 3.0, 4.0 ] ], [ [ 5.0, 6.0 ] ] ] );
+* // return <ndarray>
+*
+* var y = flattenBy( x, scale );
+* // returns <ndarray>
+*
+* var arr = ndarray2array( y );
+* // returns [ 2.0, 4.0, 6.0, 8.0, 10.0, 12.0 ]
+*/
+function flattenBy( x, options, fcn, thisArg ) {
+	var opts;
+	var ctx;
+	var cb;
+	var y;
+
+	if ( !isndarrayLike( x ) ) {
+		throw new TypeError( format( 'invalid argument. First argument must be an ndarray-like object. Value: `%s`.', x ) );
+	}
+	opts = {};
+	if ( arguments.length < 3 ) { // Case: flattenBy( x, fcn )
+		cb = options;
+	} else if ( arguments.length === 3 ) {
+		if ( isFunction( options ) ) { // Case: flattenBy( x, fcn, thisArg )
+			cb = options;
+			ctx = fcn;
+		} else { // Case: flattenBy( x, options, fcn )
+			if ( !isPlainObject( options ) ) {
+				throw new TypeError( format( 'invalid argument. Options argument must be an object. Value: `%s`.', options ) );
+			}
+			opts = options;
+			cb = fcn;
+		}
+	} else {
+		if ( !isPlainObject( options ) ) {
+			throw new TypeError( format( 'invalid argument. Options argument must be an object. Value: `%s`.', options ) );
+		}
+		opts = options;
+		cb = fcn;
+		ctx = thisArg;
+	}
+	if ( !isFunction( cb ) ) {
+		throw new TypeError( format( 'invalid argument. Callback argument must be a function. Value: `%s`.', cb ) );
+	}
+	y = map( x, cb, ctx );
+	return flatten( y, opts );
```

Comment:
You need to rethink this, as you are creating what could be a large temporary array. There is likely a more efficient way. Namely, similar to flatten, allocate a contiguous output array with the desired order, create a view over the output array which matches the input array, and then call ndarray/base/map. Then, you should be able to transform and flatten in a single call.